text/0268-acceptance-testing-refactor.md

```diff
@@ -0,0 +1,352 @@
+
+```js
+import { registerAsyncHelper } from '@ember/test';
+
+export function enableFeature(owner, featureName) {
+  let featuresService = owner.lookup('service:features');
+  featuresService.enable(featureName);
+}
+
+registerAsyncHelper('withFeature', function(app, featureName) {
+  enableFeature(app.__container__, featureName);
+});
+```
+
+This allows both the prior API (without modification) and the following:
+
+```js
+// Option 2:
+import { module, test } from 'qunit';
+import { setupAcceptanceTest } from 'ember-qunit';
+import { enableFeature } from 'addon-name-here/test-support';
+
+module('asdf', function(hooks) {
+  setupAcceptanceTest(hooks);
+
+  test('awesome test title here', function(assert) {
+    enableFeature(this.owner, 'feature-name-here');
+
+    // ...snip...
+  });
+});
+```
+
+### Registering Factory Overrides
+
+Overriding a factory is generally done to allow the test to have more control
+over the thing being tested. This is sometimes used to prevent side effects
+that are not related to the test (i.e. to prevent network calls), other times
+it is used to allow the test to inject some known state to make asserting the
+results much easier.
+
+It is currently possible to register custom factories in integration and unit
+tests, but not in acceptance tests (without using private API's that is).
+
+As of [emberjs/rfcs#232](https://github.com/emberjs/rfcs/pull/232) the
+integration/unit test API for this registration is:
+
+```js
+this.owner.register('service:stripe', MockService);
+```
+
+This RFC will allow this invocation syntax to work in all test types
+(acceptance, integration, and unit).
+
+
+## Migration
+
+It is important that both the existing acceptance testing system, and the
+newly proposed system can co-exist together. This means that new tests can be generated
+in the new style while existing tests remain untouched.
+
+However, it is likely that
+[ember-qunit-codemod](https://github.com/rwjblue/ember-qunit-codemod) will be
+able to accurately rewrite acceptance tests into the new format.
+
+# How We Teach This
+
+This change requires updates to the API documentation of `ember-qunit` and the
+main Ember guides' testing section. The changes are largely intended to reduce
+confusion, making it easier to teach and understand testing in Ember.
+
+# Drawbacks
+
+* This is a relatively large set of changes that are arguably not needed (things mostly work today).
+* One of the major hurdles in upgrading larger applications to newer Ember versions, is updating their tests to follow "new" patterns.  This RFC introduces yet another "new" thing (and proposes to deprecate the old thing), and could therefore be considered "just more churn".
+
+# Alternatives
+
+* Do nothing?
+* Make `ember-native-dom-helpers` a default addon (removing the need for DOM interaction helpers proposed here).
```

Comment:
Looks like `ember-test-helpers` with the `setupTest`, `moduleFor*` and company is already quite complicated.
Keeping DOM helpers implementation as a separate package seems like a win to me from the separation of concerns standpoint.

Reply:
@ro0gr - Forcing users to be aware of the internal packaging concerns is not ideal. From their perspective, they are using a single set of cohesive tools for testing an Ember app. Whether the implementation actually lives in `ember-test-helpers` or a new addon that is depended on by (and re-exported from) `ember-test-helpers` is not exactly important. 

Also, I disagree that `ember-test-helpers` is "too big".The non-legacy portions are actually quite small at 423 lines of code (even including the legacy code its ~ only 1353 loc).